src/realm/object-store/sync/sync_session.cpp

```diff
@@ -345,23 +401,14 @@ void SyncSession::handle_error(SyncError error)
     enum class NextStateAfterError { none, inactive, error };
     auto next_state = error.is_fatal ? NextStateAfterError::error : NextStateAfterError::none;
     auto error_code = error.error_code;
-
-    {
```

Comment:
The changes in this function are to make it avoid dropping the state mutex and then reacquiring it later, as that leaves open the opportunity for weird things to happen, and I don't think it handled state changes in all of those windows correctly.